Keep track of timestamp_field mapping as part of a data stream

martijnvg opened the pull request:
This change adds logic that keeps track of a data steam's field mapping of the timestamp_field.
When a data stream is created the mapping for the timestamp_field is extracted and stored with the data stream. Each time a rollover happens this field mapping is applied on top of the final mappings when the new backing index is created.

Having this logic is important, because the mapping inside a composable index template can change and so the mapping for the timestamp field can be changed too or removed. With this change the field mapping for the timestamp_field will stay the same for each new backing index.

martijnvg committed:
Keep track of timestamp field mapping as part of a data stream.

martijnvg committed:
use mapping source direcly instead of using mapper service to extract the relevant mapping details

martijnvg committed:
fixed rest tests

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/Data streams)

martijnvg committed:
fixed test compile errors

danhermann commented on server/src/internalClusterTest/java/org/elasticsearch/indices/DataStreamIT.java:
```diff
@@ -433,6 +444,87 @@ public void testCannotDeleteComposableTemplateUsedByDataStream() throws Exceptio
+        // Index doc that triggers creation of a data stream
+        IndexRequest indexRequest = new IndexRequest("logs-foobar").source("{}", XContentType.JSON).opType("create");
+        IndexResponse indexResponse = client().index(indexRequest).actionGet();
+        assertThat(indexResponse.getIndex(), equalTo(".ds-logs-foobar-000001"));
```
Where possible, could we use the `DataStream::getDefaultBackingIndexName` method so that any changes there are automatically picked up at least in Java classes?

danhermann commented on server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:
```diff
@@ -201,4 +204,77 @@ public boolean equals(Object o) {
         return Objects.hash(name, timeStampField, indices, generation);
     }
+
+    public static final class TimestampField implements Writeable, ToXContentObject {
```
Do we want to allow this field to contain any mapping properties or just the field's data type? Also, should we require that it contains a data type property of either `date` or `date_nanos`?

martijnvg replied:
This class should contain the mapping attributes. Part of the mapping attributes is the `type` attribute, so that data type is kept track of.

danhermann replied:
Ok, should we permit _not_ specifying a data type? A data type of either `date` or `date_nanos` is required on the composable template, so it seems like maybe it should be required on the data stream, too?

martijnvg replied:
It isn't possible that any other data type than `date` and `date_nanos` is being included the `fieldMapping` map, because no composable index template with a data stream definition can be created with timestamp field mapping that is not of type `date` or `date_nanos`.

When creating a data stream the contents of the field mapping is copied and this includes a `type` attribute, so that we know whether either `date` or `date_nanos` is specified.

I can add an assert here that checks that value for the `type` key either returns `date` or `date_nanos` for sanity checking.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateDataStreamService.java:
```diff
@@ -51,13 +55,16 @@
         new LinkedHashSet<>(List.of(DateFieldMapper.CONTENT_TYPE, DateFieldMapper.DATE_NANOS_CONTENT_TYPE));
 
     private final ClusterService clusterService;
+    private final IndicesService indicesService;
```
I think this is unused?

martijnvg replied:
A left over from an iteration that never made it into the pr.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateDataStreamService.java:
```diff
@@ -145,9 +154,18 @@ static ClusterState createDataStream(MetadataCreateIndexService metadataCreateIn
+        assert firstBackingIndex.mapping() != null : "no mapping found for backing index [" + firstBackingIndexName + "]";
+
+        String fieldName = template.getDataStreamTemplate().getTimestampField();
+        Map<String, Object> mapping = firstBackingIndex.mapping().getSourceAsMap();
```
I wonder if it was more intuitive to require that the mapping be specified on the timestamp field in the template rather than mixing it in with the index mappings. In particular, it will be more clear that those are special, not modifiable compared to the others which can change on rollover.

martijnvg replied:
I think requiring that timestamp field mapping the be specified inside the `timestamp_field` object in a composable template makes sense. It helps highlighting that the timestamp_field and its mapping are special. 

However I do think it should be made modifiable? So that new data streams being created can use an updated field mapping? This was my motivation for keeping track of the timestamp field mapping on the `DataStream` instance, so that new backing indices would use the same valid timestamp field mapping, but when new data streams are being created from the same composable template than an updated timestamp field mapping could be used.

henningandersen replied:
++, agreed, I meant unmodifiable in the sense that template changes will not reflect on already created data-streams.

martijnvg replied:
👍 I think we can change the data_stream definition in a composable template to include the mapping for the timestamp_field mapping in a seperate change?

henningandersen replied:
Yes, that can work out too.

henningandersen reviewed (commented):
I left a comment to consider about where the mapping for timestamp field is specified. Otherwise looking good.

martijnvg committed:
Merge remote-tracking branch 'es/master' into keep_track__of_ds_timestamp_field

martijnvg committed:
removed unused field

martijnvg committed:
Merge remote-tracking branch 'es/master' into keep_track__of_ds_timestamp_field

martijnvg committed:
iter

martijnvg committed:
fix versioning in serialization

martijnvg committed:
added assertions for sanity checking the type of the timestamp_field mapping

martijnvg committed:
fixed test compile and checkstyle errors

henningandersen commented on rest-api-spec/src/main/resources/rest-api-spec/test/indices.data_stream/10_basic.yml:
```diff
@@ -50,12 +50,12 @@ setup:
         name: "*"
   - match: { 0.name: simple-data-stream1 }
-  - match: { 0.timestamp_field: '@timestamp' }
+  - match: { 0.timestamp_field.field_name: '@timestamp' }
```
I wonder if this should be just `timestamp_field.name`?

I would like to also see an assert on the mapping being included and having the right content.

henningandersen commented on server/src/internalClusterTest/java/org/elasticsearch/indices/DataStreamIT.java:
```diff
@@ -433,6 +444,87 @@ public void testCannotDeleteComposableTemplateUsedByDataStream() throws Exceptio
+        assertThat(indexResponse.getIndex(), equalTo(DataStream.getDefaultBackingIndexName("logs-foobar", 2)));
+
+        // Change the template to have a different timestamp field
+        createIndexTemplate("id1", "logs-foo*", "@timestamp2");
```
I think we should rename `createIndexTemplate` to `putIndexTemplate` to signal that it also updates existing.

henningandersen commented on server/src/internalClusterTest/java/org/elasticsearch/indices/DataStreamIT.java:
```diff
@@ -433,6 +444,87 @@ public void testCannotDeleteComposableTemplateUsedByDataStream() throws Exceptio
     }
 
+    public void testChangeTimestampFieldInComposableTemplatePriorToRollOver() throws Exception {
+        createIndexTemplate("id1", "logs-foo*", "@timestamp");
```
Should we also include tests of any mapping specialization/params? AFAICS, the implementation could keep only the mapping type now and all tests would succeed.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:
```diff
@@ -493,6 +495,16 @@ private ClusterState applyCreateIndexRequestWithV2Template(final ClusterState cu
+                String mappingPath = convertFieldPathToMappingPath(dataStream.getTimeStampField().getFieldName());
+                String parentObjectFieldPath = mappingPath.substring(0, mappingPath.lastIndexOf('.'));
+                Map<String, Object> parentObjectMapper = ObjectPath.eval("_doc." + parentObjectFieldPath, mappings);
+                parentObjectMapper.put(dataStream.getTimeStampField().getFieldName(), dataStream.getTimeStampField().getFieldMapping());
```
Maybe assert that the `parentObjectMapper` does not contain a definition for the timestamp field already?

Should potentially be a production code check in case a component template can have the timestamp field too? Not sure if we guard against that.

martijnvg replied:
Currently the definition is coming from a composable index / component template. So if the name of the timestamp field mapping doesn't change then parentObjectMapper should always contain the mapping definition for the timestamp field.

Only if the name of the timestamp field changes then this logic will make sure that the original mapping definition will be used.

I think verifying whether no mapping definition exists, makes more sense when do this: https://github.com/elastic/elasticsearch/pull/58096#discussion_r440707325

In that case it is expected that the mapping definition for the timestamp_field originates not from composable index or component templates.

henningandersen replied:
Ah yes, sorry, I was one thought too far ahead.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:
```diff
@@ -201,4 +204,82 @@ public boolean equals(Object o) {
+
+        public TimestampField(String fieldName, Map<String, Object> fieldMapping) {
+            this.fieldName = fieldName;
+            this.fieldMapping = fieldMapping;
```
Should we assert that the fieldMapping contains ("type")?

martijnvg replied:
I added that assert here: https://github.com/elastic/elasticsearch/pull/58096/files#diff-82f348b800c60fabd3a34ff574e248a5R159

The reason is that in production, this is where new data stream instances are initiated.
Ideally this assertion would be added here, but then all tests need to be changes to supply: `Map.of("type, "date")` whereas currently tests supply: `Map.of()`. Maybe it makes sense to add an new constructor that just sets `fieldMapping` to `Map.of("type, "date")` and then delegates to this constructor where we can assert whether `type` key contains a valid value?

henningandersen replied:
Could we not just supply `DataStreamTestHelper.defaultMapping()` (new method) instead? A few more bytes of source code, but I would prefer asserting this here to avoid something creating this in a wrong way somewhere or under some condition.

martijnvg committed:
fixed hlrc serialization (only field name needs to get parsed)

martijnvg committed:
made field final

henningandersen reviewed (approved):
LGTM. I left a number of smaller comments to consider.

martijnvg committed:
fixed test compile errors

martijnvg committed:
rename field

martijnvg committed:
also assert mapping of timestamp_field

martijnvg committed:
renamed method

martijnvg committed:
Added integ test

martijnvg committed:
moved assertion to TimestampField class and added helper method for tests

martijnvg committed:
removed unused imports

martijnvg committed:
fixed test

martijnvg committed:
Improved logic that inserts timestamp field mapping into an mapping.
If the timestamp field path consisted out of object fields and
if the final mapping did not contain the parent field then an error
occurred, because the prior logic assumed that the object field existed.

martijnvg committed:
Merge remote-tracking branch 'es/master' into keep_track__of_ds_timestamp_field

martijnvg committed:
removed unused imports

martijnvg committed:
fixed test compile error

martijnvg committed:
Merge remote-tracking branch 'es/master' into keep_track__of_ds_timestamp_field

martijnvg committed:
fixed test compile errors after merge

martijnvg committed:
Merge remote-tracking branch 'es/master' into keep_track__of_ds_timestamp_field

jtibshirani commented:
> Having this logic is important, because the mapping inside a composable index template can change and so the mapping for the timestamp field can be changed too or removed. With this change the field mapping for the timestamp_field will stay the same for each new backing index.

I ran into this change when implementing mapping merging for composable index templates. For my context, could you explain what the goal of the change is? Why would we record the timestamp field mapping and force overwrite it instead of throwing a validation error if the new field mapping isn't okay?
